File: src/main/java/ar/com/flow/bankaccount/BankAccount.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.balance.Balance;
import ar.com.flow.bankaccount.transaction.Receipt;
import ar.com.flow.money.Money;

public interface BankAccount {
    String getCurrency();

    Balance getInitialBalance();

    Balance getBalance();

    Balance getPreviousBalance();

    void deposit(Money amountToDeposit);

    void withdraw(Money amountToWithdraw);

    void transfer(BankAccount creditAccount, Money amountToTransfer);

    void addTransactionRecord(Receipt receipt);

    Statement getStatement();

    boolean withdrawalLimitSupports(Money amount);
}

File: src/main/kotlin/ar/com/flow/bankaccount/BankAccount.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.balance.Balance
import ar.com.flow.bankaccount.transaction.receipt.Receipt
import ar.com.flow.money.Money

interface BankAccount {
    val currency: String
    val initialBalance: Balance
    val balance: Balance
    val previousBalance: Balance
    fun deposit(amountToDeposit: Money)
    fun withdraw(amountToWithdraw: Money)
    fun transfer(amountToTransfer: Money, creditAccount: BankAccount)
    fun addReceipt(receipt: Receipt)
    val statement: Statement
    fun withdrawalLimitSupports(amount: Money): Boolean
}